### ISSUE TITLE
r2l doesn't work.

### ISSUE THREAD

**Issue by linas** [opencog/opencog#2021]
This does not work:
```
(use-modules (opencog) (opencog cogserver))
(use-modules (opencog nlp))
(use-modules (opencog nlp chatbot))
(nlp-parse "look left")
```
gives the error:
```
ERROR: In procedure apply-smob/1:\nERROR: Unbound variable: pre-advmod-rule\nABORT: unbound-variable\n
```

**Comment by linas** [comment]
oh never mind. This is an old known bug in the r2l rulesbase.

**Comment by linas** [comment]
Reopening, this is a fairly serious issue.  Due to the design of the URE and how it handles rule-bases, the rules must be defined publicly, in the main guile execution environment.   This means that it is impossible to load the rules in a module -- e.g. the relex2logic module, and expect R2L to "just work".

The work-around is to have all users, even those who are NOT actually using r2l directly, to make the following magic incantation:
```
(use-modules (opencog nlp relex2logic))
(load-r2l-rulebase)
``` 

After that, other subsystems will start working, because the r2l rulebase is now in the main execution environment.

I believe that this is not specific to R2L -- I believe that all users of the URE will hit this bug.

**Comment by anitzkin** [comment]
Not to mention that the sentence "look left" is an un-grammatical colloquialism and there aren't any rules for handling it.

**Comment by linas** [comment]
Ah, right. I meant to say `look #left, #amirightyall #feelthebern`  but hit `send` by accident.

**Comment by anitzkin** [comment]
I infer that's a twitter joke, and I infer that it might be funny were I to have exposed myself to twitter.  But seriously.  "Left" is not even an adverb.  I think "look left" is a conventionalized contraction of "look to the left" -- our old friend elision mucking things up again.  Maybe you should get hypnosis so you can finally remember that solution algorithm you supposedly came up with during a boring lecture.

**Comment by linas** [comment]
Yes, I continue to try to recall that algo. But there is a deeper point: people don't talk in grammatically correct sentences.  Sometimes they don't talk at all: they hem, haw, do a chin-push and wink while filming from a selfie stick. There's a continuum of human expression, and the grammatically-correct subset of it is small (however favored it might be by those with high IQ).

Anyway, this is all very off-topic for the actual bug report.  The error message is generated for any attempted parse, not just for this particular one.

**Comment by ngeiswei** [comment]
The URE should not need the scheme  rule to be public, as long its Atomese definition is in the atomspace that is all that matters. However it is indeed convenient to use it scheme definition.

Let my try to reproduce this bug first...

**Comment by ngeiswei** [comment]
I cannot reproduce the bug, I get the following error before that

```scheme
scheme@(guile-user)> (use-modules (opencog nlp chatbot))
While compiling expression:
ERROR: no code for module (opencog nlp chatbot)
scheme@(guile-user)> (nlp-parse "look left")
;;; <stdin>:4:0: warning: possibly unbound variable `nlp-parse'
<unnamed port>:4:0: In procedure #<procedure 2bf8a00 at <current input>:4:0 ()>:
<unnamed port>:4:0: In procedure module-lookup: Unbound variable: nlp-parse
```

I have run that from the guile interpreter directly, not the cogserver. Also, I should mention that when compiling/installing opencog on my system, cmake reports that Viterbi will not be built, if that's relevant at all.

**Comment by linas** [comment]
You need to get pull req #2025 to get that to work.  Some changes somewhere broke the chattbot; that pull req fixes it.

**Comment by ngeiswei** [comment]
I still can't reproduce it. Getting the same error.

**Comment by linas** [comment]
well, `make install` will install the modules to `/usr/local/share/opencog/scm`  and so you need to have said `(add-to-load-path "/usr/local/share/opencog/scm")`  prior to the load.    The file `/usr/local/share/opencog/scm/opencog/nlp/chatbot.scm` should eist on your system, as well as a subdirectory of the same name.

**Comment by linas** [comment]
To be clear: this is a bug with the C++to-scheme interfaces for R2L. When a scheme module is being defined, it sets up a continuation, and the R2L rules are defined in that continuation. . However, when R2L actually runs, it seems to run somewhere else, and not in the continuation where the rules are defined; ergo they are not found.

Unless you happen to be good at visualizing the interplay between the scheme modules and the continuations, you'll have a hard time debugging this. 

It is possible (likely?) that this is related to bug opencog/atomspace#508 -- there, we set up scheme continuations from the C++ code, which are incomplete, because the C++ code does not set everything up.   Perhaps we should stop doing this setup in the C++ code ....

**Comment by ngeiswei** [comment]
`chatbot.scm` is not installed on my system (even after running `make install`), and I don't find anywhere in the CMake config files a command that would install it.

**Comment by linas** [comment]
```
ADD_GUILE_MODULE (chatbot.scm
   chat-utils.scm
   process-query.scm
)
```

The `ADD_GUILE_MODULE` is some code that Amen worte, it should be doing the install; its doing it for me.

**Comment by linas** [comment]
lib/OpenCogFunctions.cmake

**Comment by linas** [comment]
Closing. I think I found it.  if one uses the sock `opencog.conf` files when starting the cogserver, then it ends up clobbering the load of all the guile modules, totally messing things up.   Thus, all sorts of crazy, hard-to-track down bugs that "dont make sense" start happening.    The cogserver conf files need to be trimmed back.

**Comment by linas** [comment]
Closing fixed by #2075
